### DISCUSSION TITLE
I tried to use the source code to build clangd, but I was confused

### DISCUSSION THREAD

**Discussion by Yurii-huang** [clangd/clangd#1352] [Q&A] [answered]
I see in the LLVM parameter that can be used `-DLLVM_ENABLE_PROJECTS`, it seems that there is no guidance document on how to build from source code.

**Comment by Yurii-huang** [comment]
> I see in the LLVM parameter that can be used `-DLLVM_ENABLE_PROJECTS`, it seems that there is no guidance document on how to build from source code.

I'll record how I built the source code:

1. fork remote git repos from llvm
2. Cd to git repos dir and execute cmd :
`mkdir build && cd build &&  cmake -S ../llvm -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra" -DCMAKE_INSTALL_PREFIX=./install -DCMAKE_BUILD_TYPE=release && make -j8 `
3. Wait for compilation to complete

**Comment by HighCommander4** [comment, marked as answer]
See https://clangd.llvm.org/faq#how-do-i-build-clangd-from-sources.

I think the main thing missing from your command is that `LLVM_ENABLE_PROJECTS` needs to include `clang-tools-extra` as well, i.e. `-DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra"`

**Reply by Yurii-huang** [reply]
I try add `clang-tools-extra` params, but not working, log:
```bash
$ uos@uos-PC:~/funning/llvm-project-llvmorg-13.0.0/build$ cmake -S ../llvm -DLLVM_ENABLE_PROJECTS=clang;clang-tools-extra -DCMAKE_INSTALL_PREFIX=./install -DCMAKE_BUILD_TYPE=release
-- Version: 0.0.0
-- Performing Test HAVE_GNU_POSIX_REGEX -- failed to compile
-- Performing Test HAVE_POSIX_REGEX -- success
-- Performing Test HAVE_STEADY_CLOCK -- success
-- Configuring done
-- Generating done
-- Build files have been written to: /home/uos/funning/llvm-project-llvmorg-13.0.0/build
-bash: clang-tools-extra：not found command
```

**Reply by Yurii-huang** [reply]
![image](https://user-images.githubusercontent.com/34232854/198211443-8e63ad57-b2f3-405f-9793-f602e7fe2ee4.png)

**Reply by HighCommander4** [reply]
Sorry forgot that it needs to be quoted, `-DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra"`, because `;` has special meaning in the shell syntax (it's a command separator).

**Reply by Yurii-huang** [reply]
LOL!
Yes, I didn't find this problem either.

**Reply by Yurii-huang** [reply]
So build from source can to use params:
```bash
"-DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra"
```
